### ISSUE TITLE
tksupport does not allow application quit on MacOS X

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3766]
|<img alt="rowen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rowen reported|
|-|-|
|Trac ID|trac#3766|
|Type|defect|
|Created|2009-04-08 22:39:58Z|

On MacOS X Aqua Tcl/Tk puts up a standard Quit menu item that seems to be hard-wired to do something inside Tcl/Tk itself (perhaps quit the mainloop)? This action is not noticed by Twisted, which keeps the application running as if nothing happened.

Is there some way that the tksupport reactor could notice that this action has occurred and handle it correctly? Perhaps this is intrinsic to having Twisted run the tk event loop instead of the other way around, in which case it may be unsoluble from the Twisted side.

I have not found any way to override the behavior of the Quit menu item in Tcl/Tk (though perhaps the lack of that is a Tcl/Tk bug and I do plan to request it of that project).

<details><summary>Searchable metadata</summary>

```
trac-id__3766 3766
type__defect defect
reporter__rowen rowen
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1239230398000000 1239230398000000
changetime__1239373108000000 1239373108000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="rowen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|rowen set status to `closed`|
|-|-|

Just after posting this Kevin Walzer provided an answer. The Quit menu on MacOS X calls the tcl "exit" function. One can easily intercept this in Python/Tkinter using createcommand, e.g.:
   root.createcommand('exit', myQuitFunction)

**Comment by twisted-trac** [comment]
|<img alt="rowen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|rowen commented|
|-|-|

It turns out there is a more official technique, courtesy of Daniel A. Steffen:

### begin quote

if the tcl proc '::tk::mac::Quit' is defined, it gets called instead of exit upon receipt of a 'quit' apple event (which is
what the standard quit menu item sends, as well as other entities in the system such as the dock menu quit item or the graceful Finder restart/shutdown, this is why the quit menu item is special). If ::tk::mac::Quit returns false, the application is not quit.

other default apple event handlers are also defined, c.f.
tkMacOSXHLEvent.c

### end quote

In other words: root.createcommand("::tk::mac::Quit", myQuitFunction)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Does this mean that the WM_DELETE_WINDOW event doesn't happen on OS X?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
